### ISSUE TITLE
OPENSTEP 4.2 can't boot from dc390/amdTitle

### ISSUE THREAD

**Issue by Randrianasulu** [86Box/86Box#4956]
### What happened?

Installer can't find cdrom

### Configuration file

```ini
[General]
vid_renderer = qt_opengl3
host_cpu = AMD FX(tm)-4300 Quad-Core Processor
uuid = 763f958e-d3d5-54c6-b20a-f8792674e83a
sound_gain = 18
dpi_scale = 0

[Machine]
machine = tx97
cpu_family = pentium_p54c
cpu_speed = 200000000
cpu_multi = 3
cpu_use_dynarec = 1
fpu_softfloat = 0
time_sync = local
fpu_type = internal
mem_size = 262144

[Video]
gfxcard = n9_9fx_771_pci

[Input devices]
mouse_type = logibus

[Sound]
fm_driver = nuked
sndcard = sb16

[Network]
net_01_link = 0
net_02_link = 0
net_03_link = 0
net_04_link = 0
net_01_card = pcnetpci
net_01_net_type = slirp

[Storage controllers]
hdc = none
cassette_mode = load
scsicard_1 = am53c974

[Floppy and CD-ROM drives]
cdrom_01_parameters = 1, scsi
cdrom_01_scsi_location = 0:01
fdd_01_type = 35_2hd
fdd_02_type = none
fdd_01_turbo = 1
fdd_01_check_bpb = 0
fdd_01_image_history_01 = /dev/shm/FLOPPY144.img
cdrom_01_image_history_01 = /home/ftp/incoming/sw_openstep42.iso
cdrom_01_image_history_02 = /home/ftp/incoming/ansib_bench.iso
cdrom_01_image_history_03 = /dev/shm/OpenStep 4.2 Developer Tools CD By JPS.iso
cdrom_01_type = 86BOX_CD-ROM_1.00
fdd_01_fn = /dev/shm/FLOPPY144-2.img
cdrom_01_image_path = /dev/shm/OPENSTEP_4.2.iso

[Novell NE2000 #1]
base = 0300
irq = 3
mac = e6:cc:b0
bios_addr = 00000

[Adaptec AHA-154xCP #1]
bios_rev = v1_02_en

[Tekram DC-390 PCI #1]
bios = 1

[BusLogic BT-958D PCI #1]
bios = 1

[Hard disks]
hdd_01_parameters = 63, 16, 4063, 0, scsi
hdd_01_scsi_location = 0:00
hdd_01_fn = /dev/shm/2gb_os42.img
hdd_01_speed = 2000_7200rpm

[Matrox Millennium II]
memory = 4

[S3 Vision968 PCI (Diamond Stealth64 Video VRAM)]
memory = 4

[S3 ViRGE (Diamond Stealth 3D 2000) PCI]
memory = 2
bilinear = 1
dithering = 1

[S3 ViRGE/DX (Diamond Stealth 3D 2000 Pro) PCI]
memory = 4
bilinear = 1
dithering = 1

[S3 ViRGE (325) PCI]
memory = 2
bilinear = 1
dithering = 1

[Tseng Labs ET4000/w32p Rev. D PCI (Diamond Stealth32)]
memory = 2

[Logitech/Microsoft Bus Mouse]
base = 023C
irq = 3
hz = 60
buttons = 2

[AMD PCnet-FAST III #1]
mac = e6:77:19

[AMD PCnet-PCI II #1]
mac = ad:14:3d

[S3 Vision964 PCI (Diamond Stealth64 VRAM)]
memory = 2

[S3 Vision968 PCI (Number 9 9FX 771)]
memory = 4
```

### Operating system

Linux

### CPU

AMD FX(tm)-4300 Quad-Core Processor

### 86Box version

git commit a3ad407a21ae75f6b8db2c63d81930e75c4ce5ff

### Build architecture

Linux - x86 (32-bit)

### Build type

- [X] New recompiler
- [ ] Debug build

### Download source

Official website (Jenkins, GitHub)

### Additional context

![86box-amd-PCscsi-problem](https://github.com/user-attachments/assets/33ff605e-d8b4-418d-811d-eaf281bd6ba8)

same error with Tekram DC390

It works with BusLogic BT958D

**Comment by TC1995** [comment]
Not sure how I can fix this to be honest.

**Comment by Randrianasulu** [comment]
add  more debug? :)

I tried my Linux live cd based on 2.4.37.11, it boots fine on ide-attached cdrom and then I can modprobe tmscsim (kernel module) and it sees 86Box hdd (11 mb test image I created just for that experiment), but reads via hdparm -Tt /dev/sda tend to be extremely slow, like 80 kilobytes/sec! may be just disk was too small for hdparm  version I have in there ....

https://archive.org/details/bitsavers_amddataShe_5075364
datasheet, hopefully.

**Comment by TC1995** [comment]
I already know about the datasheet.

**Comment by TC1995** [comment]
Also, in the case of NeXTSTEP/OPENSTEP, it issues a transfer count of 0 in that exact instance.

**Comment by Randrianasulu** [comment]
![openstep_33_amd_driver_scrinshoot](https://github.com/user-attachments/assets/5ce0eb06-e65a-48d5-8cd4-ba1a6a629c14)


floppy image apparently in (ufs?) 4.3 format, so mounting on Linux side fails ..

On installed OPENSTEP 4.2 side I can see driver's internals. including this Default.table (parameters)

**Comment by TC1995** [comment]
I meant that in the code of the driver of NeXTSTEP/OPENSTEP, it somehow gets a transfer count of 0, thus, failing to get a proper size for SCSI devices attached to the controller.

**Comment by Randrianasulu** [comment]
I am looking at pdf ... there is start transfer count reg
at 
b+00h
b+04h
b+38h

so yeah, three-byte register ....

is it responsible for ) you see?

Is there way to log pci access to this specific device in 86Box?

**Comment by TC1995** [comment]
that, exactly, it incorrectly gets set to 0.

**Comment by Randrianasulu** [comment]
```
The least significant byte is located at address ((B)+OOh). the middle byte is located at address ((B)+04h). and the most significant byte is located at address ((B)+38h). Register ((B)+38h) extends the total width of the register from 16 to 24 bits. and is only enabled when the Enable Features bit (bit 6) of Control Register Two is set to a value of '1'. This sets the maximum transfer count to 16 MBytes. When a value of '0' is written to these registers. the transfer count will be set to the maximum.
```

may be not incorrectly?

**Comment by TC1995** [comment]
I tried that in the past, didn't work. Though I may need to try again.

**Comment by Randrianasulu** [comment]
also, you can mount NeXTSTEP 3.3 floppy with
mount 3.3_Addl_Drivers.img /mnt/test -t ufs -o ufstype=nextstep

under Linux.

and see drivers, while they compiled with Objective-c, so probably not most useful.

there was pdf on writing drivers for NeXTSTEP/OPENSTEP, but it said to be incomplete
https://wiki.preterhuman.net/images/f/ff/DriverKit.pdf 

I think this

**Comment by TC1995** [comment]
I'm not on a Linux host.

**Comment by TC1995** [comment]
Also, the real deal would be DMA_STC (b+44h), which is also 0 there.

**Comment by Randrianasulu** [comment]
![Screenshot_20241111-022054](https://github.com/user-attachments/assets/8d9eb6c6-923e-4d14-87d8-b87ce8e9e626)


may be I misread this, but  I think it says that b+44h should be loaded with same val as those three, yet those 3 can be set to 0 for max transfer

**Comment by TC1995** [comment]
yes, and now I discovered it enables MDL Scatter-Gather.

**Comment by Randrianasulu** [comment]
[4.2DriverKit.tar.gz](https://github.com/user-attachments/files/17700334/4.2DriverKit.tar.gz)

driverkit examples, including AMD PCSCSI driver example with rtf readme ....

**Comment by TC1995** [comment]
Ok, this also happens on QEMU as well, but I'm not sure if it's really a problem of the driver or the implementation used on both emulators.

**Comment by OBattler** [comment]
Yes, we inherited this from QEMU, which does not implement the MDL Scatter-Gather - I'm going to have to implement that at some point.

**Comment by TC1995** [comment]
Well, I believe it's not even MDL Scatter Gather, because the TC (Transfer Count) is still 0 from the esp side (non-PCI) of the DC390/53c974 PCI.

**Comment by TC1995** [comment]
Ok nvm, it's indeed the MDL S/G as it's issued right during Inquiry (SCSI command 0x12).

**Comment by TC1995** [comment]
This is now fixed, wait for it to be merged and the binary built.
